content/IA-update_connections/connection-state-and-recovery.textile

```diff
@@ -0,0 +1,292 @@
+
+```[jsall]
+  /* remove a listener registered for a single event */
+  realtime.connection.off('connected', myListener);  
+
+  /* remove a listener registered for all events */
+  realtime.connection.off(myListener);  
+
+  /* remove all event listeners */
+  realtime.connection.off();
+```
+
+```[java]
+  /* remove a single listener */
+  realtime.connection.off(myListener);
+
+  /* remove all event listeners */
+  realtime.connection.off();
+```
+
+```[csharp]
+  /* remove a single handler */
+  realtime.Connection.Off(action);
+
+  /* remove all event handlers */
+  realtime.Connection.Off();
+```
+
+```[ruby]
+  # remove a listener registered for a single even
+  realtime.connection.off :connected, &block
+
+  # remove a listener registered for all events
+  realtime.connection.off &block
+
+  # remove all event listeners
+  realtime.connection.off
+```
+
+```[objc]
+  // remove a listener registered for a single event
+  [realtime.connection off:ARTRealtimeConnectionEventConnected listener:listener];
+
+  // remove a listener registered for all events
+  [realtime.connection off:listener];
+
+  // remove all event listeners
+  [realtime.connection off];
+```
+
+```[swift]
+  // remove a listener registered for a single event
+  realtime.connection.off(.connected, listener: listener)
+
+  // remove a listener registered for all events
+  realtime.connection.off(listener)
+
+  // remove all event listeners
+  realtime.connection.off()
+```
+
+h3(#handling-failures). Handling failures
+
+The client libraries will attempt to automatically recover from non-fatal error conditions. However, it will emit events to say what it's doing, so you can handle them yourself if you prefer.
+
+h4(#fatal-errors). Fatal errors
+
+Some classes of errors are fatal. These cause the connection to move to the @FAILED@ state. The client library will not attempt any automatic recovery actions. For example, if your token expires and the client library has no way to get a new token (so no authUrl and authCallback), the connection will enter the @FAILED@ state
+
+While the library will not automatically attempt to reconnect in the @FAILED@ state, explicit calls to "@connect()@":/api/realtime-sdk/connection#connect will make the client try again.
+
+h4(#nonfatal-errors). Nonfatal errors
+
+Other classes of error are nonfatal. For example, a client may have network connectivity issues. The library will attempt to automatically reconnect and recover from these sort of issues, as detailed in the @DISCONNECTED@ and @SUSPENDED@ explanations in the "Available connection states":#connection-states section.
+
+If message continuity is lost in the process, e.g. because you have been disconnected from Ably for more than two minutes, the library will notify you though the @resumed@-flag mechanism, detailed in the "Channels and Messages":/realtime/channels#nonfatal-errors page.
+
+h3(#connection-state-recovery). Connection state recovery
```

Comment:
I think this should be an h2 and located before the section on handling connection failures.

Reply:
Done.